noxfile.py

```diff
@@ -0,0 +1,228 @@
+import pathlib
+import re
+from typing import Optional, List
+import nox
+
+nox.options.reuse_existing_virtualenvs = True
+nox.options.sessions = [
+    "black",
+    "pylint",
+    "mypy",
+    "check_yaml",
+    "check_json",
+    "check_toml",
+    "check_eof",
+    "check_trailing_space",
+    "check_lf",
+]
+
+
+@nox.session
+def black(session: nox.Session):
+    """Checks formatting linting and typing"""
+    session.install("-r", "requirements/dev-black.txt")
+    session.run("black", ".")
+
+
+@nox.session(python=False)
+def pylint(session: nox.Session):
+    """Run pylint using in default env"""
+    session.run("pip", "install", "-r", "requirements/dev-pylint.txt")
+    session.run("pylint", "sphinx_immaterial")
+
+
+@nox.session(python=False)
+def mypy(session: nox.Session):
```

Comment:
I believe mypy results are Python version-dependent, so we should do this for all Python versions.

Reply:
> I thought there might be a way to cache nox venvs without having any of the deps (which are set to use latest versions) in the cache, but I didn't look into it enough.

Nope. Anything installed while using a nox venv is installed to the venv. `session.run("pip install")` doesn't use the system env if using a nox venv. Note, `session.install()` can't be used if using the system env (when `@nox.session(python=False)`). So, we'll just have the CI ignore the nox venvs for mypy and pylint (only caching the black and pre-commit-hooks venvs).

PS - I think it would be overkill to run pylint and mypy with different versions of sphinx (like the unit tests are now doing).